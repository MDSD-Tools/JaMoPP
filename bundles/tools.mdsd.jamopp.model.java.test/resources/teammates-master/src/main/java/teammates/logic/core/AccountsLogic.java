package teammates.logic.core;

import java.util.List;

import teammates.common.datatransfer.attributes.AccountAttributes;
import teammates.common.datatransfer.attributes.InstructorAttributes;
import teammates.common.datatransfer.attributes.StudentAttributes;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InstructorUpdateException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.api.AccountsDb;

/**
 * Handles operations related to accounts.
 *
 * @see AccountAttributes
 * @see AccountsDb
 */
public final class AccountsLogic {

    private static final AccountsLogic instance = new AccountsLogic();

    private final AccountsDb accountsDb = AccountsDb.inst();

    private ProfilesLogic profilesLogic;
    private CoursesLogic coursesLogic;
    private InstructorsLogic instructorsLogic;
    private StudentsLogic studentsLogic;

    private AccountsLogic() {
        // prevent initialization
    }

    public static AccountsLogic inst() {
        return instance;
    }

    void initLogicDependencies() {
        profilesLogic = ProfilesLogic.inst();
        coursesLogic = CoursesLogic.inst();
        instructorsLogic = InstructorsLogic.inst();
        studentsLogic = StudentsLogic.inst();
    }

    /**
     * Creates an account.
     *
     * @return the created account
     * @throws InvalidParametersException if the account is not valid
     * @throws EntityAlreadyExistsException if the account already exists in the database.
     */
    AccountAttributes createAccount(AccountAttributes accountData)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return accountsDb.createEntity(accountData);
    }

    /**
     * Gets an account.
     */
    public AccountAttributes getAccount(String googleId) {
        return accountsDb.getAccount(googleId);
    }

    /**
     * Returns true if the given account exists and is an instructor.
     */
    public boolean isAccountAnInstructor(String googleId) {
        AccountAttributes a = accountsDb.getAccount(googleId);
        return a != null && a.isInstructor();
    }

    /**
     * Joins the user as a student.
     */
    public StudentAttributes joinCourseForStudent(String registrationKey, String googleId)
            throws InvalidParametersException, EntityDoesNotExistException, EntityAlreadyExistsException {
        StudentAttributes student = validateStudentJoinRequest(registrationKey, googleId);

        // Register the student
        student.setGoogleId(googleId);
        try {
            studentsLogic.updateStudentCascade(
                    StudentAttributes.updateOptionsBuilder(student.getCourse(), student.getEmail())
                            .withGoogleId(student.getGoogleId())
                            .build());
        } catch (EntityDoesNotExistException e) {
            assert false : "Student disappeared while trying to register";
        }

        if (accountsDb.getAccount(googleId) == null) {
            createStudentAccount(student);
        }

        return student;
    }

    /**
     * Joins the user as an instructor and sets the institute if it is not null.
     * If the given institute is null, the instructor is given the institute of an existing instructor of the same course.
     */
    public InstructorAttributes joinCourseForInstructor(String key, String googleId)
            throws InvalidParametersException, EntityDoesNotExistException, EntityAlreadyExistsException {
        InstructorAttributes instructor = validateInstructorJoinRequest(key, googleId);

        // Register the instructor
        instructor.setGoogleId(googleId);
        try {
            instructorsLogic.updateInstructorByEmail(
                    InstructorAttributes.updateOptionsWithEmailBuilder(instructor.getCourseId(), instructor.getEmail())
                            .withGoogleId(instructor.getGoogleId())
                            .build());
        } catch (EntityDoesNotExistException e) {
            assert false : "Instructor disappeared while trying to register";
        } catch (InstructorUpdateException e) {
            assert false;
        }

        AccountAttributes account = accountsDb.getAccount(googleId);
        String instituteToSave = coursesLogic.getCourseInstitute(instructor.getCourseId());

        if (account == null) {
            try {
                createAccount(AccountAttributes.builder(googleId)
                        .withName(instructor.getName())
                        .withEmail(instructor.getEmail())
                        .withInstitute(instituteToSave)
                        .withIsInstructor(true)
                        .build());
            } catch (EntityAlreadyExistsException e) {
                assert false : "Account already exists.";
            }
        } else {
            makeAccountInstructor(googleId);
        }

        // Update the googleId of the student entity for the instructor which was created from sample data.
        StudentAttributes student = studentsLogic.getStudentForEmail(instructor.getCourseId(), instructor.getEmail());
        if (student != null) {
            student.setGoogleId(googleId);
            studentsLogic.updateStudentCascade(
                    StudentAttributes.updateOptionsBuilder(student.getCourse(), student.getEmail())
                            .withGoogleId(student.getGoogleId())
                            .build());
        }

        return instructor;
    }

    private InstructorAttributes validateInstructorJoinRequest(String registrationKey, String googleId)
            throws EntityDoesNotExistException, EntityAlreadyExistsException {
        InstructorAttributes instructorForKey = instructorsLogic.getInstructorForRegistrationKey(registrationKey);

        if (instructorForKey == null) {
            throw new EntityDoesNotExistException("No instructor with given registration key: " + registrationKey);
        }

        if (instructorForKey.isRegistered()) {
            if (instructorForKey.getGoogleId().equals(googleId)) {
                AccountAttributes existingAccount = accountsDb.getAccount(googleId);
                if (existingAccount != null && existingAccount.isInstructor()) {
                    throw new EntityAlreadyExistsException("Instructor has already joined course");
                }
            } else {
                throw new EntityAlreadyExistsException("Instructor has already joined course");
            }
        } else {
            // Check if this Google ID has already joined this course
            InstructorAttributes existingInstructor =
                    instructorsLogic.getInstructorForGoogleId(instructorForKey.getCourseId(), googleId);

            if (existingInstructor != null) {
                throw new EntityAlreadyExistsException("Instructor has already joined course");
            }
        }

        return instructorForKey;
    }

    private StudentAttributes validateStudentJoinRequest(String registrationKey, String googleId)
            throws EntityDoesNotExistException, EntityAlreadyExistsException {

        StudentAttributes studentRole = studentsLogic.getStudentForRegistrationKey(registrationKey);

        if (studentRole == null) {
            throw new EntityDoesNotExistException("No student with given registration key: " + registrationKey);
        }

        if (studentRole.isRegistered()) {
            throw new EntityAlreadyExistsException("Student has already joined course");
        }

        // Check if this Google ID has already joined this course
        StudentAttributes existingStudent =
                studentsLogic.getStudentForCourseIdAndGoogleId(studentRole.getCourse(), googleId);

        if (existingStudent != null) {
            throw new EntityAlreadyExistsException("Student has already joined course");
        }

        return studentRole;
    }

    /**
     * Downgrades an instructor account to student account.
     *
     * <p>Cascade deletes all instructors associated with the account.
     */
    public void downgradeInstructorToStudentCascade(String googleId) throws EntityDoesNotExistException {
        instructorsLogic.deleteInstructorsForGoogleIdCascade(googleId);

        try {
            accountsDb.updateAccount(
                    AccountAttributes.updateOptionsBuilder(googleId)
                            .withIsInstructor(false)
                            .build()
            );
        } catch (InvalidParametersException e) {
            assert false : "Invalid account data detected unexpectedly "
                    + "while removing instruction privileges from account " + googleId + ": " + e.getMessage();
        }
    }

    /**
     * Makes an account as an instructor account.
     */
    void makeAccountInstructor(String googleId) throws InvalidParametersException, EntityDoesNotExistException {
        accountsDb.updateAccount(
                AccountAttributes.updateOptionsBuilder(googleId)
                        .withIsInstructor(true)
                        .build()
        );
    }

    /**
     * Deletes both instructor and student privileges, as long as the account and associated student profile.
     *
     * <ul>
     * <li>Fails silently if no such account.</li>
     * </ul>
     */
    public void deleteAccountCascade(String googleId) {
        if (accountsDb.getAccount(googleId) == null) {
            return;
        }

        profilesLogic.deleteStudentProfile(googleId);

        // to prevent orphan course
        List<InstructorAttributes> instructorsToDelete =
                instructorsLogic.getInstructorsForGoogleId(googleId, false);
        for (InstructorAttributes instructorToDelete : instructorsToDelete) {
            if (instructorsLogic.getInstructorsForCourse(instructorToDelete.getCourseId()).size() <= 1) {
                // the instructor is the last instructor in the course
                coursesLogic.deleteCourseCascade(instructorToDelete.getCourseId());
            }
        }

        instructorsLogic.deleteInstructorsForGoogleIdCascade(googleId);
        studentsLogic.deleteStudentsForGoogleIdCascade(googleId);
        accountsDb.deleteAccount(googleId);
    }

    /**
     * Creates a student account.
     */
    private void createStudentAccount(StudentAttributes student)
            throws InvalidParametersException, EntityAlreadyExistsException {

        AccountAttributes account = AccountAttributes.builder(student.getGoogleId())
                .withEmail(student.getEmail())
                .withName(student.getName())
                .withIsInstructor(false)
                .withInstitute(coursesLogic.getCourseInstitute(student.getCourse()))
                .build();

        accountsDb.createEntity(account);
    }

}
