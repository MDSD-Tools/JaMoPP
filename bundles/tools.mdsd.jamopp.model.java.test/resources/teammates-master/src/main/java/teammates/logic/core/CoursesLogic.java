package teammates.logic.core;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import teammates.common.datatransfer.AttributesDeletionQuery;
import teammates.common.datatransfer.InstructorPrivileges;
import teammates.common.datatransfer.attributes.AccountAttributes;
import teammates.common.datatransfer.attributes.CourseAttributes;
import teammates.common.datatransfer.attributes.InstructorAttributes;
import teammates.common.datatransfer.attributes.StudentAttributes;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.Const;
import teammates.common.util.Logger;
import teammates.storage.api.CoursesDb;

/**
 * Handles operations related to courses.
 *
 * @see CourseAttributes
 * @see CoursesDb
 */
public final class CoursesLogic {

    private static final Logger log = Logger.getLogger();

    private static final CoursesLogic instance = new CoursesLogic();

    /* Explanation: This class depends on CoursesDb class but no other *Db classes.
     * That is because reading/writing entities from/to the database is the
     * responsibility of the matching *Logic class.
     * However, this class can talk to other *Logic classes. That is because
     * the logic related to one entity type can involve the logic related to
     * other entity types.
     */

    private final CoursesDb coursesDb = CoursesDb.inst();

    private AccountsLogic accountsLogic;
    private FeedbackSessionsLogic feedbackSessionsLogic;
    private FeedbackQuestionsLogic fqLogic;
    private FeedbackResponsesLogic frLogic;
    private FeedbackResponseCommentsLogic frcLogic;
    private InstructorsLogic instructorsLogic;
    private StudentsLogic studentsLogic;

    private CoursesLogic() {
        // prevent initialization
    }

    public static CoursesLogic inst() {
        return instance;
    }

    void initLogicDependencies() {
        accountsLogic = AccountsLogic.inst();
        feedbackSessionsLogic = FeedbackSessionsLogic.inst();
        fqLogic = FeedbackQuestionsLogic.inst();
        frLogic = FeedbackResponsesLogic.inst();
        frcLogic = FeedbackResponseCommentsLogic.inst();
        instructorsLogic = InstructorsLogic.inst();
        studentsLogic = StudentsLogic.inst();
    }

    /**
     * Gets the institute associated with the course.
     */
    public String getCourseInstitute(String courseId) {
        CourseAttributes cd = getCourse(courseId);
        assert cd != null : "Trying to getCourseInstitute for inexistent course with id " + courseId;
        return cd.getInstitute();
    }

    /**
     * Creates a course.
     *
     * @return the created course
     * @throws InvalidParametersException if the course is not valid
     * @throws EntityAlreadyExistsException if the course already exists in the database.
     */
    CourseAttributes createCourse(CourseAttributes courseToCreate)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return coursesDb.createEntity(courseToCreate);
    }

    /**
     * Creates a course and an associated instructor for the course.
     *
     * <br/>Preconditions: <br/>
     * * {@code instructorGoogleId} already has an account and instructor privileges.
     */
    public void createCourseAndInstructor(String instructorGoogleId, CourseAttributes courseToCreate)
            throws InvalidParametersException, EntityAlreadyExistsException {

        AccountAttributes courseCreator = accountsLogic.getAccount(instructorGoogleId);
        assert courseCreator != null : "Trying to create a course for a non-existent instructor :" + instructorGoogleId;
        assert courseCreator.isInstructor()
                : "Trying to create a course for a person who doesn't have instructor privileges :" + instructorGoogleId;

        CourseAttributes createdCourse = createCourse(courseToCreate);

        // Create the initial instructor for the course
        InstructorPrivileges privileges = new InstructorPrivileges(
                Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_COOWNER);
        InstructorAttributes instructor = InstructorAttributes
                .builder(createdCourse.getId(), courseCreator.getEmail())
                .withName(courseCreator.getName())
                .withGoogleId(instructorGoogleId)
                .withPrivileges(privileges)
                .build();

        try {
            instructorsLogic.createInstructor(instructor);
        } catch (EntityAlreadyExistsException | InvalidParametersException e) {
            // roll back the transaction
            coursesDb.deleteCourse(createdCourse.getId());
            String errorMessage = "Unexpected exception while trying to create instructor for a new course "
                                  + System.lineSeparator() + instructor.toString();
            assert false : errorMessage;
        }
    }

    /**
     * Gets the course with the specified ID.
     */
    public CourseAttributes getCourse(String courseId) {
        return coursesDb.getCourse(courseId);
    }

    /**
     * Returns true if the course with ID courseId is present.
     */
    public boolean isCoursePresent(String courseId) {
        return coursesDb.getCourse(courseId) != null;
    }

    /**
     * Used to trigger an {@link EntityDoesNotExistException} if the course is not present.
     */
    void verifyCourseIsPresent(String courseId) throws EntityDoesNotExistException {
        if (!isCoursePresent(courseId)) {
            throw new EntityDoesNotExistException("Course does not exist: " + courseId);
        }
    }

    /**
     * Returns a list of section names for the course with valid ID courseId.
     *
     * @param courseId Course ID of the course
     */
    public List<String> getSectionsNameForCourse(String courseId) throws EntityDoesNotExistException {
        verifyCourseIsPresent(courseId);

        List<StudentAttributes> studentDataList = studentsLogic.getStudentsForCourse(courseId);

        Set<String> sectionNameSet = new HashSet<>();
        for (StudentAttributes sd : studentDataList) {
            if (!sd.getSection().equals(Const.DEFAULT_SECTION)) {
                sectionNameSet.add(sd.getSection());
            }
        }

        List<String> sectionNameList = new ArrayList<>(sectionNameSet);
        sectionNameList.sort(null);

        return sectionNameList;
    }

    /**
     * Returns team names for a particular courseId.
     *
     * <p>Note: This method does not returns any Loner information presently.
     * Loner information must be returned as we decide to support loners in future.
     */
    public List<String> getTeamsForCourse(String courseId) throws EntityDoesNotExistException {

        if (getCourse(courseId) == null) {
            throw new EntityDoesNotExistException("The course " + courseId + " does not exist");
        }

        return studentsLogic.getStudentsForCourse(courseId)
                .stream()
                .map(StudentAttributes::getTeam)
                .distinct()
                .sorted()
                .collect(Collectors.toList());
    }

    /**
     * Returns a list of {@link CourseAttributes} for all courses a given student is enrolled in.
     *
     * @param googleId The Google ID of the student
     */
    public List<CourseAttributes> getCoursesForStudentAccount(String googleId) {
        List<StudentAttributes> studentDataList = studentsLogic.getStudentsForGoogleId(googleId);

        List<String> courseIds = studentDataList.stream()
                .filter(student -> !getCourse(student.getCourse()).isCourseDeleted())
                .map(StudentAttributes::getCourse)
                .collect(Collectors.toList());

        return coursesDb.getCourses(courseIds);
    }

    /**
     * Returns a list of {@link CourseAttributes} for all courses for a given list of instructors
     * except for courses in Recycle Bin.
     */
    public List<CourseAttributes> getCoursesForInstructor(List<InstructorAttributes> instructorList) {
        assert instructorList != null;

        List<String> courseIdList = instructorList.stream()
                .filter(instructor -> !coursesDb.getCourse(instructor.getCourseId()).isCourseDeleted())
                .map(InstructorAttributes::getCourseId)
                .collect(Collectors.toList());

        List<CourseAttributes> courseList = coursesDb.getCourses(courseIdList);

        // Check that all courseIds queried returned a course.
        if (courseIdList.size() > courseList.size()) {
            for (CourseAttributes ca : courseList) {
                courseIdList.remove(ca.getId());
            }
            log.severe("Course(s) was deleted but the instructor still exists: " + System.lineSeparator()
                    + courseIdList.toString());
        }

        return courseList;
    }

    /**
     * Returns a list of {@link CourseAttributes} for soft-deleted courses for a given list of instructors.
     */
    public List<CourseAttributes> getSoftDeletedCoursesForInstructors(List<InstructorAttributes> instructorList) {
        assert instructorList != null;

        List<String> softDeletedCourseIdList = instructorList.stream()
                .filter(instructor -> coursesDb.getCourse(instructor.getCourseId()).isCourseDeleted())
                .map(InstructorAttributes::getCourseId)
                .collect(Collectors.toList());

        List<CourseAttributes> softDeletedCourseList = coursesDb.getCourses(softDeletedCourseIdList);

        if (softDeletedCourseIdList.size() > softDeletedCourseList.size()) {
            for (CourseAttributes ca : softDeletedCourseList) {
                softDeletedCourseIdList.remove(ca.getId());
            }
            log.severe("Course(s) was deleted but the instructor still exists: " + System.lineSeparator()
                    + softDeletedCourseIdList.toString());
        }

        return softDeletedCourseList;
    }

    /**
     * Updates a course by {@link CourseAttributes.UpdateOptions}.
     *
     * <p>If the {@code timezone} of the course is changed, cascade the change to its corresponding feedback sessions.
     *
     * @return updated course
     * @throws InvalidParametersException if attributes to update are not valid
     * @throws EntityDoesNotExistException if the course cannot be found
     */
    public CourseAttributes updateCourseCascade(CourseAttributes.UpdateOptions updateOptions)
            throws InvalidParametersException, EntityDoesNotExistException {
        CourseAttributes oldCourse = coursesDb.getCourse(updateOptions.getCourseId());
        CourseAttributes updatedCourse = coursesDb.updateCourse(updateOptions);

        if (!updatedCourse.getTimeZone().equals(oldCourse.getTimeZone())) {
            feedbackSessionsLogic
                    .updateFeedbackSessionsTimeZoneForCourse(updatedCourse.getId(), updatedCourse.getTimeZone());
        }

        return updatedCourse;
    }

    /**
     * Deletes a course cascade its students, instructors, sessions, responses and comments.
     *
     * <p>Fails silently if no such course.
     */
    public void deleteCourseCascade(String courseId) {
        if (getCourse(courseId) == null) {
            return;
        }

        AttributesDeletionQuery query = AttributesDeletionQuery.builder()
                .withCourseId(courseId)
                .build();
        frcLogic.deleteFeedbackResponseComments(query);
        frLogic.deleteFeedbackResponses(query);
        fqLogic.deleteFeedbackQuestions(query);
        feedbackSessionsLogic.deleteFeedbackSessions(query);
        studentsLogic.deleteStudents(query);
        instructorsLogic.deleteInstructors(query);

        coursesDb.deleteCourse(courseId);
    }

    /**
     * Moves a course to Recycle Bin by its given corresponding ID.
     * @return the time when the course is moved to the recycle bin
     */
    public Instant moveCourseToRecycleBin(String courseId) throws EntityDoesNotExistException {

        return coursesDb.softDeleteCourse(courseId);
    }

    /**
     * Restores a course from Recycle Bin by its given corresponding ID.
     */
    public void restoreCourseFromRecycleBin(String courseId) throws EntityDoesNotExistException {
        coursesDb.restoreDeletedCourse(courseId);
    }

}
