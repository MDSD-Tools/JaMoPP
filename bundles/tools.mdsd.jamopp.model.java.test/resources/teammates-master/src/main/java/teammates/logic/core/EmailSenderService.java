package teammates.logic.core;

import teammates.common.exception.EmailSendingException;
import teammates.common.util.EmailSendingStatus;
import teammates.common.util.EmailWrapper;

/**
 * An email sender interface used by services for sending emails.
 */
public interface EmailSenderService {

    /**
     * Parses the {@code wrapper} email object to specific implementations of email object
     * used by the service.
     */
    Object parseToEmail(EmailWrapper wrapper);

    /**
     * Sends the email packaged as a {@code wrapper}.
     */
    EmailSendingStatus sendEmail(EmailWrapper wrapper) throws EmailSendingException;

}
