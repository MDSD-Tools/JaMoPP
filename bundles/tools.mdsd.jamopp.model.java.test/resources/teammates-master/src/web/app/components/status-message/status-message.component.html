<div class="alert alert-{{ statusMessage.color }}" *ngFor="let statusMessage of messages">
  <button class="close" (click)="dismiss(statusMessage)">
    <span aria-hidden="true">&times;</span>
  </button>
  <div style="padding-right: 30px;" [innerHtml]="statusMessage.message"></div>
</div>
