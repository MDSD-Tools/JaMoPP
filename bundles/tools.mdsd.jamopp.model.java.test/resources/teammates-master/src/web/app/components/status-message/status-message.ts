/**
 * Represents a status message and a severity color.
 *
 * The color is adapted from Bootstrap's color scheme.
 */
export interface StatusMessage {
  message: string;
  color: string;
}
