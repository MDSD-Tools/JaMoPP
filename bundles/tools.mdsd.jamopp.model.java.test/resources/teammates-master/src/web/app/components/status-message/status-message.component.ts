import { Component, Input } from '@angular/core';
import { StatusMessage } from './status-message';

/**
 * List of status messages.
 */
@Component({
  selector: 'tm-status-message',
  templateUrl: './status-message.component.html',
  styleUrls: ['./status-message.component.scss'],
})
export class StatusMessageComponent {

  @Input() messages: StatusMessage[] = [];

  constructor() { }

  /**
   * Dismisses the status message.
   */
  dismiss(message: StatusMessage): void {
    this.messages.splice(this.messages.indexOf(message), 1);
  }

}
