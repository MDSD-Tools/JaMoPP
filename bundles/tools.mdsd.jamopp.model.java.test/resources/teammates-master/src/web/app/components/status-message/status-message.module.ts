import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { StatusMessageComponent } from './status-message.component';

/**
 * Module for status messages.
 */
@NgModule({
  imports: [
    CommonModule,
  ],
  exports: [
    StatusMessageComponent,
  ],
  declarations: [
    StatusMessageComponent,
  ],
})
export class StatusMessageModule { }
