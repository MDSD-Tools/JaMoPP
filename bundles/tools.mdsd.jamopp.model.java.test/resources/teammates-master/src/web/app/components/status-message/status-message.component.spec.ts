import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { StatusMessage } from './status-message';
import { StatusMessageComponent } from './status-message.component';

describe('StatusMessageComponent', () => {
  let component: StatusMessageComponent;
  let fixture: ComponentFixture<StatusMessageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [StatusMessageComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(StatusMessageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  const messages: StatusMessage[] = [
    { message: 'a', color: 'black' },
    { message: 'b', color: 'red' },
    { message: 'c', color: 'green' },
  ];

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should show all three messages', () => {
    component.messages = messages;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should delete a message after clicking button', () => {
    component.messages = messages;
    fixture.detectChanges();

    const buttons: HTMLElement[] = fixture.nativeElement.querySelectorAll('div.alert button');

    buttons[0].click();
    fixture.detectChanges();

    const statusMessagesAfterClick: HTMLElement[] = fixture.nativeElement.querySelectorAll('div.alert div');
    const messageAfterDelete: StatusMessage[] = [
      { message: 'b', color: 'red' },
      { message: 'c', color: 'green' },
    ];

    expect(component.messages).toEqual(messageAfterDelete);
    expect(statusMessagesAfterClick.length).toEqual(2);
    expect(statusMessagesAfterClick[0].innerHTML).toEqual('b');
    expect(statusMessagesAfterClick[1].innerHTML).toEqual('c');

    buttons[1].click();
    buttons[2].click();
    fixture.detectChanges();

    const deleteAll: HTMLElement[] = fixture.nativeElement.querySelectorAll('div.alert div');
    const messageAfterDeleteAll: StatusMessage[] = [];

    expect(component.messages).toEqual(messageAfterDeleteAll);

    expect(deleteAll.length).toEqual(0);
  });

});
