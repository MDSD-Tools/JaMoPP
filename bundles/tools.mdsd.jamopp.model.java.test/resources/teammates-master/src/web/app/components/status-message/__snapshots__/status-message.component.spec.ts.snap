// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusMessageComponent should show all three messages 1`] = `
<tm-status-message
  messages={[Function Array]}
>
  <div
    class="alert alert-black"
  >
    <button
      class="close"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    <div
      style="padding-right: 30px;"
    >
      a
    </div>
  </div><div
    class="alert alert-red"
  >
    <button
      class="close"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    <div
      style="padding-right: 30px;"
    >
      b
    </div>
  </div><div
    class="alert alert-green"
  >
    <button
      class="close"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    <div
      style="padding-right: 30px;"
    >
      c
    </div>
  </div>
</tm-status-message>
`;

exports[`StatusMessageComponent should snap with default view 1`] = `
<tm-status-message
  messages={[Function Array]}
>
  
</tm-status-message>
`;
